Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm

```diff
@@ -460,9 +464,88 @@
     return WebCore::Cursor::Type::Pointer;
 }
 
-auto UnifiedPDFPlugin::pdfElementTypesForPluginPoint(const WebCore::IntPoint&) const -> PDFElementTypes
+WebCore::IntPoint UnifiedPDFPlugin::pointInDocumentSpaceForPluginPoint(const WebCore::IntPoint& point) const
 {
-    return { };
+    return WebCore::AffineTransform { }.scale(1.0f / m_documentLayout.scale()).translate(m_scrollOffset).mapPoint(point);
+}
+
+std::optional<PDFDocumentLayout::PageIndex> UnifiedPDFPlugin::nearestPageIndexForDocumentPoint(const WebCore::IntPoint& point) const
+{
+    std::optional<PDFDocumentLayout::PageIndex> nearestPageIndex;
+
+    switch (m_documentLayout.displayMode()) {
+    case PDFDocumentLayout::DisplayMode::TwoUp:
+    case PDFDocumentLayout::DisplayMode::TwoUpContinuous:
+        // FIXME: Implement page index resolution for TwoUp display modes.
+        notImplemented();
+        break;
+    case PDFDocumentLayout::DisplayMode::SinglePage:
+        // There is only one page to be near.
+        return 0;
+    case PDFDocumentLayout::DisplayMode::Continuous: {
+        // Walk down the document, find the first page bound that contains the position.
+        // Note that we don't consider display direction (vertical vs horizontal).
+        for (PDFDocumentLayout::PageIndex index = 0; index < m_documentLayout.pageCount(); ++index) {
+            auto pageBounds = m_documentLayout.boundsForPageAtIndex(index);
+            pageBounds.expand(PDFDocumentLayout::documentMargin);
+            pageBounds.contract(PDFDocumentLayout::pageMargin);
+            // At this point, this page index necessarily corresponds to the
+            // nearest page to the point (so far), so we always keep track of it.
+            nearestPageIndex = index;
+            if (point.y() <= pageBounds.maxY())
+                break;
```

Comment:
Is it possible to ever *not* hit this condition? If so `nearestPageIndex` would point to the last page and would that be correct in that scenario or would `std::nullopt` be more appropriate?

Reply:
Addressed in [af10ddc](https://github.com/WebKit/WebKit/commit/af10ddcd0e5fdcb907e4c6dc388e54411872b172)